services/head_tracker.go

```diff
@@ -175,73 +181,114 @@ func (ht *HeadTracker) onNewHead(head *models.BlockHeader) {
 	}
 }
 
-func (ht *HeadTracker) subscribeToNewHeads(headers chan models.BlockHeader) (models.EthSubscription, error) {
-	sub, err := ht.store.TxManager.SubscribeToNewHeads(headers)
-	if err != nil {
-		return nil, err
+func (ht *HeadTracker) listenForNewHeads() {
+	defer ht.listenForNewHeadsWg.Done()
+	defer ht.unsubscribeFromHead()
+
+	for {
+		ht.resubscriptionPoll()
+
+	SubscriptionLoop:
+		for {
+			select {
+			case <-ht.done:
+				return
+			case header, open := <-ht.headers:
+				if !open {
+					break SubscriptionLoop
+				}
+				number := header.ToIndexableBlockNumber()
+				logger.Debugw(fmt.Sprintf("Received header %v with hash %s", presenters.FriendlyBigInt(number.ToInt()), header.Hash().String()), "hash", header.Hash())
+				if err := ht.Save(number); err != nil {
+					logger.Error(err.Error())
+				} else {
+					ht.onNewHead(&header)
+				}
+			case err, open := <-ht.headSubscription.Err():
+				if open && err != nil {
+					logger.Errorw("Error in new head subscription, disconnected", "err", err)
+					break SubscriptionLoop
+				}
+			}
+		}
 	}
-	go func() {
-		err := <-sub.Err()
-		if err != nil {
-			logger.Errorw("Error in new head subscription, disconnected", "err", err)
-			ht.reconnectLoop()
+}
+
+// resubscriptionPoll periodically attempts to connect to the ethereum node via websocket.
+// It returns true on success, and false if cut short by a done request and did not connect.
+func (ht *HeadTracker) resubscriptionPoll() {
+	ht.sleeper.Reset()
+	for {
+		ht.unsubscribeFromHead()
+		logger.Info("Connecting to node ", ht.store.Config.EthereumURL, " in ", ht.sleeper.Duration())
+		select {
+		case <-ht.done:
+			return
+		case <-time.After(ht.sleeper.After()):
+			err := ht.subscribeToHead()
+			if err != nil {
+				logger.Errorw(fmt.Sprintf("Error connecting to %v", ht.store.Config.EthereumURL), "err", err)
+			} else {
+				logger.Info("Connected to node ", ht.store.Config.EthereumURL)
+				select {
```

Comment:
Is this some kind of "poll a channel" pattern? Could we put this in a function to document its behavior, otherwise looks a bit "clever".

Reply:
This is how you do a non blocking send in go to an unbuffered channel. Should I place it in function or comment?

Reply:
This is how you do a non blocking send in go to an unbuffered channel. Should I place it in function or comment?